### PR TITLE
Kserve V2 - Fix for Bytes output

### PR DESCRIPTION
**Pull request by shrinath-suresh**
Signed-off-by: Shrinath Suresh <shrinath@ideas2it.com>

## Description

The current kserve v2 service envelope throws up the following error when the input and output are sent as bytes.

```
2022-01-24T15:02:51,749 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG - Traceback (most recent call last):
2022-01-24T15:02:51,749 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/ts/service.py", line 102, in predict
2022-01-24T15:02:51,749 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     ret = self._entry_point(input_batch, self.context)
2022-01-24T15:02:51,750 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/ts/torch_handler/request_envelope/base.py", line 31, in handle
2022-01-24T15:02:51,750 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     results = self.format_output(results)
2022-01-24T15:02:51,750 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/ts/torch_handler/request_envelope/kservev2.py", line 133, in format_output
2022-01-24T15:02:51,750 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     response["outputs"] = self._batch_to_json(data)
2022-01-24T15:02:51,750 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/ts/torch_handler/request_envelope/kservev2.py", line 142, in _batch_to_json
2022-01-24T15:02:51,750 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     output.append(self._to_json(item))
2022-01-24T15:02:51,750 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/ts/torch_handler/request_envelope/kservev2.py", line 155, in _to_json
2022-01-24T15:02:51,751 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     output_data["datatype"] = _to_datatype(data_ndarray.dtype)
2022-01-24T15:02:51,751 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -   File "/home/ubuntu/anaconda3/lib/python3.8/site-packages/ts/torch_handler/request_envelope/kservev2.py", line 41, in _to_datatype
2022-01-24T15:02:51,751 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG -     datatype = _NumpyToDatatype[as_str]
2022-01-24T15:02:51,751 [INFO ] W-9000-bert_1.0-stdout MODEL_LOG - KeyError: '<U12'
2022-01-24T15:02:51,751 [INFO ] W-9000-bert_1.0-stdout MODEL_METRICS - HandlerTime.Milliseconds:572.8|#ModelName:bert,Level:Model|#hostname:ip-172-31-18-40,requestID:0c0ff862-429d-47ed-92e7-cbd667ac7740,timestamp:1643036571
```

**Steps for reproducing the above error**

1. Install TS from source 
```
git clone https://github.com/pytorch/serve.git
cd serve
python ts_scripts/install_dependencies.py
python ts_scripts/install_from_src.py
```

2. Install kserve using pip

```
pip install kserve
```

3. Download dependent packages for bert example
```
cd serve/examples/Huggingface_Transformers
Set `captum_explanation` to `false` in `setup_config.json`
python Download_Transformer_models.py
```

4. Generate mar file
```
torch-model-archiver --model-name BERTSeqClassification --version 1.0 --serialized-file Transformer_model/pytorch_model.bin --handler ./Transformer_handler_generalized.py --extra-files "Transformer_model/config.json,./setup_config.json,./Seq_classification_artifacts/index_to_name.json"
```

5. Move the mar file to model-store folder
```
sudo mv BERTSeqClassification.mar /mnt/models/model-store/
```

6. Create the following config.properties file in `/mnt/models/config/config.properties`
```
inference_address=http://0.0.0.0:8085
management_address=http://0.0.0.0:8085
metrics_address=http://0.0.0.0:8082
enable_envvars_config=true
install_py_dep_per_model=true
enable_metrics_api=true
service_envelope=kservev2
metrics_format=prometheus
NUM_WORKERS=1
number_of_netty_threads=4
job_queue_size=10
model_store=/mnt/models/model-store
model_snapshot={"name":"startup.cfg","modelCount":1,"models":{"bert":{"1.0":{"defaultVersion":true,"marName":"BERTSeqClassification.mar","minWorkers":1,"maxWorkers":5,"batchSize":1,"maxBatchDelay":5000,"responseTimeout":120}}}}
```

7. Start torchserve and wait for model to load
```
torchserve --start --ts-config /mnt/models/config/config.properties --ncs
```

8. Start kserve
```
cd serve/kubernetes/kserve/kserve_wrapper
python __main__.py
```
9. Create a new input file `bert_sc_bytes.json` with the following contents
```
{
  "id": "d3b15cad-50a2-4eaf-80ce-8b0a428bd298",
  "inputs": [{
    "name": "4b7c7d4a-51e4-43c8-af61-04639f6ef4bc",
    "shape": -1,
    "datatype": "BYTES",
    "data": "this year business is good"
  }]
}
```
10. Run the following curl command

```
curl -v -H "ContentType: application/json" http://localhost:8080/v2/models/bert/infer -d @./bert_sc_bytes.json
```

**Expected Result:**

```
{"id": "d3b15cad-50a2-4eaf-80ce-8b0a428bd298", "model_name": "BERTSeqClassification", "model_version": "1.0", "outputs": [{"name": "predict", "shape": [], "datatype": "BYTES", "data": ["Not Accepted"]}]}
```

**Actual Result:**

``` 
Error -  KeyError: '<U12'
 ```

Fixes #(issue)

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Feature/Issue validation/testing

Please describe the tests [UT/IT] that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced. 
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- UT/IT execution results

- Logs

## Checklist:

- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [x] New and existing unit tests pass locally with these changes?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

### CODE REVIEW

File: ts/torch_handler/request_envelope/kservev2.py [hunk 1]
```diff
@@ -29,6 +29,7 @@
 
 # NOTE: numpy has more types than v2 protocol
 _NumpyToDatatype["object"] = "BYTES"
+_NumpyToDatatype["U"] = "BYTES"
```
Comment (yuzisun) [inline ts/torch_handler/request_envelope/kservev2.py:32]: What does “U” mean for numpy?

Comment (shrinath-suresh) [reply ts/torch_handler/request_envelope/kservev2.py:32]: U stands for "Unicode String" - https://www.w3schools.com/python/numpy/numpy_data_types.asp

Comment (yuzisun) [reply ts/torch_handler/request_envelope/kservev2.py:32]: good to leave a comment here with the reference?

Comment (shrinath-suresh) [reply ts/torch_handler/request_envelope/kservev2.py:35]: why not !! Added the comment

**Comment by lxning** [comment]
@shrinath-suresh could you post test log?

**Comment by shrinath-suresh** [comment]
> @shrinath-suresh could you post test log?

do you mean torchserve log ? sure i can retest and upload it

**Comment by shrinath-suresh** [comment]
> > @shrinath-suresh could you post test log?
> 
> do you mean torchserve log ? sure i can retest and upload it

Please find the logs attached

[bytest_ts_error_log.log](https://github.com/pytorch/serve/files/8005442/bytest_ts_error_log.log)
[BertBytesFixSuccess.log](https://github.com/pytorch/serve/files/8005495/BertBytesFixSuccess.log)

**Review by HamidShojanazeri — APPROVED** [review]
Thanks @shrinath-suresh, LGTM.

**Review by lxning — APPROVED** [review]

**Pull request merged**